clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp

```diff
@@ -679,77 +710,34 @@ clang::dpct::BarrierFenceSpaceAnalyzer::analyze(const CallExpr *CE,
 }
 
 bool clang::dpct::BarrierFenceSpaceAnalyzer::hasOverlappingAccessAmongWorkItems(
-    int KernelDim, const DeclRefExpr *DRE) {
+    int KernelCallBlockDim, const DeclRefExpr *DRE) {
   using namespace ast_matchers;
-  if (KernelDim != 1) {
+  if (KernelCallBlockDim != 1) {
     return true;
   }
 
   const ArraySubscriptExpr *ASE = getArraySubscriptExpr(DRE);
   if (!ASE)
     return true;
-
-  const Expr *InitExpr = getIdxOfASEConstValueExpr(ASE);
-  if (!InitExpr)
+  auto Res = getIdxOfASEConstValueExpr(ASE);
+  if (!std::get<0>(Res))
     return true;
 
-  // Check if Index variable match pattern: blockIdx.x * blockDim.x +
-  // threadIdx.x
-  // Case 1: blockIdx.x * blockDim.x + threadIdx.x
-  // Case 2: blockDim.x * blockIdx.x + threadIdx.x
-  // Case 3: threadIdx.x + blockIdx.x * blockDim.x
-  // Case 4: threadIdx.x + blockDim.x * blockIdx.x
-  const BinaryOperator *BOAdd = dyn_cast<BinaryOperator>(InitExpr);
-  if (!BOAdd || BOAdd->getOpcode() != BinaryOperatorKind::BO_Add)
-    return true;
-  const BinaryOperator *BOMul = dyn_cast<BinaryOperator>(BOAdd->getLHS());
-  if (BOMul &&
-      isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOAdd->getRHS()),
-                                 "threadIdx", "__fetch_builtin_x")) {
-    if (isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getLHS()),
-                                   "blockIdx", "__fetch_builtin_x") &&
-        isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getRHS()),
-                                   "blockDim", "__fetch_builtin_x")) {
-      // Case 1
-      return false;
-    }
-    if (isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getRHS()),
-                                   "blockIdx", "__fetch_builtin_x") &&
-        isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getLHS()),
-                                   "blockDim", "__fetch_builtin_x")) {
-      // Case 2
-      return false;
-    }
-    return true;
-  }
-  BOMul = dyn_cast<BinaryOperator>(BOAdd->getRHS());
-  if (BOMul &&
-      isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOAdd->getLHS()),
-                                 "threadIdx", "__fetch_builtin_x")) {
-    if (isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getLHS()),
-                                   "blockIdx", "__fetch_builtin_x") &&
-        isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getRHS()),
-                                   "blockDim", "__fetch_builtin_x")) {
-      // Case 3
-      return false;
-    }
-    if (isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getRHS()),
-                                   "blockIdx", "__fetch_builtin_x") &&
-        isIterationSpaceBuiltinVar(dyn_cast<PseudoObjectExpr>(BOMul->getLHS()),
-                                   "blockDim", "__fetch_builtin_x")) {
-      // Case 4
-      return false;
-    }
-    return true;
+  IndexAnalysis SMA(std::get<0>(Res));
```

Comment:
Please update the name of SMA? and add comment of the analysis logic.

Reply:
done